File: api/lib/utils/color.mjs
import { clamp, round } from "./math.mjs";

export function hexToRgb(hex) {
  const normalized = String(hex).replace("#", "").trim();
  const expanded =
    normalized.length === 3
      ? normalized
          .split("")
          .map((part) => `${part}${part}`)
          .join("")
      : normalized;

  if (!/^[\da-fA-F]{6}$/.test(expanded)) {
    return null;
  }

  return {
    r: Number.parseInt(expanded.slice(0, 2), 16),
    g: Number.parseInt(expanded.slice(2, 4), 16),
    b: Number.parseInt(expanded.slice(4, 6), 16),
  };
}

export function mixHexColors(baseHex, mixHex, weight) {
  const base = hexToRgb(baseHex);
  const mix = hexToRgb(mixHex);

  if (!base || !mix) {
    return baseHex;
  }

  const ratio = clamp(weight, 0, 1);
  const toHex = (value) =>
    Math.round(value).toString(16).padStart(2, "0");

  return `#${toHex(base.r + (mix.r - base.r) * ratio)}${toHex(base.g + (mix.g - base.g) * ratio)}${toHex(base.b + (mix.b - base.b) * ratio)}`;
}

export function toRgba(hex, alpha) {
  const color = hexToRgb(hex);

  if (!color) {
    return `rgba(255, 255, 255, ${round(alpha)})`;
  }

  return `rgba(${color.r}, ${color.g}, ${color.b}, ${round(alpha)})`;
}

File: internal/util/color.go
package util

import (
	"fmt"
	"strconv"
	"strings"
)

type RGB struct {
	R int
	G int
	B int
}

func HexToRGB(hex string) *RGB {
	normalized := strings.TrimSpace(strings.TrimPrefix(hex, "#"))
	expanded := normalized
	if len(normalized) == 3 {
		var builder strings.Builder
		builder.Grow(6)
		for _, part := range normalized {
			builder.WriteRune(part)
			builder.WriteRune(part)
		}
		expanded = builder.String()
	}

	if len(expanded) != 6 {
		return nil
	}

	for _, character := range expanded {
		switch {
		case character >= '0' && character <= '9':
		case character >= 'a' && character <= 'f':
		case character >= 'A' && character <= 'F':
		default:
			return nil
		}
	}

	r, _ := strconv.ParseInt(expanded[0:2], 16, 64)
	g, _ := strconv.ParseInt(expanded[2:4], 16, 64)
	b, _ := strconv.ParseInt(expanded[4:6], 16, 64)

	return &RGB{R: int(r), G: int(g), B: int(b)}
}

func MixHexColors(baseHex, mixHex string, weight float64) string {
	base := HexToRGB(baseHex)
	mix := HexToRGB(mixHex)
	if base == nil || mix == nil {
		return baseHex
	}

	ratio := ClampFloat(weight, 0, 1)
	toHex := func(value float64) string {
		return fmt.Sprintf("%02x", int(mathRound(value)))
	}

	return "#" +
		toHex(float64(base.R)+(float64(mix.R-base.R)*ratio)) +
		toHex(float64(base.G)+(float64(mix.G-base.G)*ratio)) +
		toHex(float64(base.B)+(float64(mix.B-base.B)*ratio))
}

func ToRGBA(hex string, alpha float64) string {
	color := HexToRGB(hex)
	if color == nil {
		return "rgba(255, 255, 255, " + NumberString(Round(alpha)) + ")"
	}

	return fmt.Sprintf(
		"rgba(%d, %d, %d, %s)",
		color.R,
		color.G,
		color.B,
		NumberString(Round(alpha)),
	)
}

func mathRound(value float64) float64 {
	if value < 0 {
		return float64(int(value - 0.5))
	}

	return float64(int(value + 0.5))
}
